### DISCUSSION TITLE
Drop Starlette

### DISCUSSION THREAD

**Discussion by Goldziher** [litestar-org/litestar#612] [Ideas]
We reached a point where we can relatively easily drop Starlette as a dependency - we use only a few pieces of it's ASGI kit we can implement on our own:

1. The Response classes.

2. The ASGI LifeSpan logic

3. CORs Middleware

4. GZip Middleware

5. Allowed Hosts Middleware

6. The Test client

And the exported http statuses . 

I propose we tackle these one at a time and when we are ready, drop Starlette. This can be done with almost no breaking changes and just a few adjustments to exports. 

Please give your opinions on this

**Comment by provinzkraut** [comment]
I'm curious, could you elaborate the motivation behind dropping starlette? Is it currently restricting starlite development in some way?

**Reply by Goldziher** [reply]
Sure. 

So first off - we don't control the Starlette code base, which means breaking changes can happen without our discretion. This happened with the switch of the test client to `httpx`, which while pretty cool, also posed some difficulties for us.

Secondly, we no longer need Starlette because we actually implemented the complex parts on our own - we have a completely different routing level. We have our own requests, we also use a different typing (asgiref based etc.) which is no longer compatible.

Finally, the parts of Starlette we do use, we cannot change or optimize and are bottlenecks for us. I would much rather have our own response classes and the few middleware we have from there, and have the ability to modify and extend them into our patterns. 

From my PoV the only real reason to keep Starlette around is that it gives Starlite more credibility. But Starlite is more established now and we have a significant community. I think we are reaching a point were this is no longer required, and in fact it works against us- Starlite is not merely a "wrapper" around Starlette. It is in fact about 2 to 3 times larger and significantly more complex.

**Reply by provinzkraut** [reply]
Thanks for the explanation! That makes a lot of sense actually!

**Reply by v3ss0n** [reply]
I think if we are moving away from starlette , we should also look into
- AnyIO
- Async task groups

**Reply by provinzkraut** [reply]
@v3ss0n Can you elaborate?

Do you suggest we drop those or adopt them more?

**Comment by provinzkraut** [comment]
How would you go about deprecating like the CORS middleware for example? Users might rely on the starlette implementation. I assume something like implementing a version that's backwards compatible with starlette first and issuing a `PendingDeprecationWarning` that tells users to use starlite's middleware, and only allow backwards-incompatible changes in a future major version?

**Reply by peterschutt** [reply]
Our public interface for CORS is the `CORSConfig` object and our use of the starlette `CORSMiddleware` is really an implementation detail and isn't documented anywhere. I'd therefore argue that we don't really need to deprecate our internal use of it.

However, we've advertised a level of compatibility with that middleware pattern and that shouldn't change just because we are not using starlette internally. So users should still be able to use that implementation if they choose by passing it to `middleware`.

**Comment by peterschutt** [comment]
Also, `ServerErrorMiddleware` for debug responses.

**Comment by provinzkraut** [comment]
There's a few other things currently being used from starlette. I did a little search with the `ast` module, and these are all the things imported from starlette (excluding individual status codes because of verbosity)

- **starlette.datastructures**
  - MutableHeaders (6 imports)
  - URL (4 imports)
  - Headers (5 imports)
  - Address (2 imports)
  - URLPath (1 import)
  - ImmutableMultiDict (1 import)
  - State (1 import)
- **starlette.middleware**
  - Middleware (3 imports)
  - cors.CORSMiddleware (2 imports)
  - trustedhost.TrustedHostMiddleware (2 imports)
  - errors.ServerErrorMiddleware (1 import)
  - gzip.GZipMiddleware (1 import)
  - gzip.GZipResponder (1 import)
  - base.BaseHTTPMiddleware (2 imports)
  - base.RequestResponseEndpoint (1 import)
- **starlette.staticfiles**
  - StaticFiles (2 imports)
- **starlette.responses**
  - Response (7 imports)
  - RedirectResponse (3 imports)
  - FileResponse (2 imports)
  - StreamingResponse (2 imports)
  - PlainTextResponse (2 imports)
  - HTMLResponse (1 import)
  - JSONResponse (1 import)
- **starlette.routing**
  - Router (1 import)
  - get_name (5 imports)
- **starlette.exceptions**
  - HTTPException (5 imports)
- **starlette.requests**
  - cookie_parser (2 imports)
  - SERVER_PUSH_HEADERS_TO_COPY (1 import)
- **starlette.testclient**
  - TestClient (3 imports)
- **starlette.background**
  - BackgroundTask (1 import)
  - BackgroundTasks (1 import)
- **starlette.websockets**
  - WebSocketDisconnect (4 imports)


So it's largely middleware, datastructures and responses. Maybe we should open a tracking issue for this to have an overview of the progress?

**Reply by peterschutt** [reply]
Nice stuff. Tracking issue sounds good, or multiple issues and we can track them together in a project, up to you.

**Reply by provinzkraut** [reply]
Hm. Project might be too much complexity for this imho. How about we create a tracking issue for overall progress + a milestone so we can group issues / PRs concerning this together?

**Reply by peterschutt** [reply]
Perfect.

**Comment by provinzkraut** [comment]
Is there any reason in particular to define all the status codes ourselves and not use the stdlib [http.HTTPStatus](https://docs.python.org/3/library/http.html#http.HTTPStatus)? It's missing the websocket codes, but that's easy enough to add.

**Reply by Goldziher** [reply]
three reasons:

1. the stdlib offers an enum where each entry is a tuple, what we need are simple constants.
2. the stdlib has different names then those used in our codebase and those used by our users, so this will be a big breaking change.
3. we want to improve typing, which is what I did - check it out, and the stdlib isn't fitting for this.

**Comment by Goldziher** [comment]
Regarding the `TestClient`, this is going to be the biggest hurdle for us because:

1. its a big chunk of rather complex code.
2. its quite strongly coupled to Starlette, which makes it even now problematic for us because we diverged with our websocket implementation.

My suggestion is to create a new package called `starlite-test-client`, which will include a version "forked" from Starlette. It wont be an actual github fork, but will include a modified version of the Starlette code. We will then move to using this package.

Thoughts?

**Reply by provinzkraut** [reply]
If we're going that route, it should be included as a non-optional dependency in starlite IMO. Having a built-in test client is a rather essential feature.

But why can't we include this "fork" in the starlite codebase?

**Reply by Goldziher** [reply]
We can include it in the starlite code base. We currently allow testing to be optional - so that was the thinking. I would also consider it a builtin, but we require a bunch of other things as well for testing.

**Reply by provinzkraut** [reply]
IMO testing as advertised currently is a bit contradictory anyway.

>Testing is a first class citizen in Starlite, which offers several powerful testing utilities **out of the box**

>Testing utils rely on extra dependencies. So make sure to install them

If they need to be installed, it's not really out of the box. But if we're going to keep this as is, then I agree that it makes sense to put the testing client in a separate package.

**Reply by Tobi-De** [reply]
> If they need to be installed, it's not really out of the box. But if we're going to keep this as is, then I agree that it makes sense to put the testing client in a separate package.
It could be a separate package required and installed directly with starlite

**Reply by provinzkraut** [reply]
Issue: #668

**Reply by ubernostrum** [reply]
I would really like for Starlite to go back to not requiring the `testing` extra in order to be able to run tests -- as noted, it makes testing feel much less like a "first class" thing, and when the `httpx` switch happened it actually broke testing flows for a Starlite project of mine that was just installing `starlite` and not `starlite[testing]`, which was annoying.

**Reply by Goldziher** [reply]
we are keeping it inside starlite - as for not being optional, the problem is that it requires HTTPX, and our helpers for sessions require cryptography.

**Reply by ubernostrum** [reply]
I personally don't mind if plain `starlite` has additional dependencies in order to support a testing client and the other tools. I *do* mind that I have to install `starlite[testing]` to be able to use the testing tools. Like the other comment said, it feels weird when the claim in the documentation is that testing is available "out of the box". So whatever the required set of dependencies is, I'd hope they become part of the default Starlite dependencies, and I can just `pip install starlite` without having to remember that that's "wrong" for any project with tests (which is all of my projects).

**Reply by peterschutt** [reply]
> I do mind that I have to install starlite[testing] to be able to use the testing tools.

Extras are ubiquitous in python packaging.

FYI, #130 is the issue that was the genesis of moving the test client dependencies out of the main install.

Whether we include testing dependencies in with the library dependencies or not, some people will like it, some will hate it, and some will accept it for what it is and not worry about it. I for one prefer to keep it the way it is, we have enough churn and I agree with the premise of #130

**Reply by ubernostrum** [reply]
I'm familiar with extras in Python packaging.

But if Starlite's documentation is going to claim testing is "first class" with "powerful testing utilities out of the box", then the very next line of the documentation should not contradict that by reminding you that actually you need to specify optional extras in order to use those testing utilities.

And honestly, tools like test clients have been table stakes for Python web frameworks for so many years now, that it feels strange to have to go to extra effort to get access to them. Even Flask, which is about as minimal as frameworks get, bundles a test client in the default install.

**Reply by provinzkraut** [reply]
I already commented on this, I just want to add that if we're keeping it the way it is (requiring install with `starlite[testing]`), we should change the wording in our documentation, because then it's not as advertised - it's not "out of the box". It fits in the box, it works with the box, the box knows about it, but it's not in there by default.

**Reply by Goldziher** [reply]
I have no objection to having this part of the core install. @peterschutt ? @cofin ? @infohash ?

**Reply by infohash** [reply]
In my view, starlite is growing bigger in terms of its features, so, we can keep independent components of starlite as optional to install. It's not a one time thing, this standard should be applicable on any independent components in future versions as well to keep consistency in distributing starlite. I think `starlite[testing]` is fine as it's not a required component to run starlite.

Users additionally have to install testing frameworks to write or run tests anyway so `starlite[testing]` should be treated as an extra instead of out of box.

**Reply by peterschutt** [reply]
I think we should leave developers with as much control over their production dependencies as possible. From my experience, I usually have httpx explicitly installed anyway, but cryptography is simply extra MBs and therefore unnecessary cost.

**Reply by Goldziher** [reply]
ok, so lets update the docs.

**Reply by Goldziher** [reply]
One other option occurred to me-

We can make httpx and the testing client a builtin, and only move the other testing dependencies into extra by doing some isolation of methods. 

What do you think? I think that it might be preferable - I guess most users will use testing, but a relative few will create test sessions, which require cryptography.

**Reply by ubernostrum** [reply]
I would be fine with that. An easy, out-of-the-box test client is crucial, and a web framework without one really just feels incomplete.

**Reply by provinzkraut** [reply]
@Goldziher Superb idea!

> I guess most users will use testing, but a relative few will create test sessions, which require cryptography.

Add to that the fact that the server-side sessions I’m working on don’t require cryptography, and users could just use the dummy memory backend for testing unless they’re doing integration tests for cookies / storage!

This would effectively make cryptography not a dependency of the `SessionMiddleware` and therefore of the test client, but only for specific session backends.

**Reply by peterschutt** [reply]
> We can make httpx and the testing client a builtin, and only move the other testing dependencies into extra by doing some isolation of methods.

As mentioned, for me it is rare not to have httpx available anyway - although that's coincidental to the fact that httpx is my preferred async http client library. For others it may not be, and we'd be shoehorning users into using httpx, or to have to choose to have two http client libraries in prod dependencies. I suppose now would be a good time for anyone who fits in that group to speak out if they are around and care enough to weigh in..

At the very least having cryptography as optional draws a line in the sand, so that any further niche testing dependencies we might want to acquire along the way are treated in the same manner.

So, my preference is still to keep the status quo with optional testing extras and to amend the docs, but I won't push back if the general consensus is to take this half-way approach.

**Reply by Goldziher** [reply]
@provinzkraut when you're back on your feet, can you update what needs to be done here vis a vis the testing client now that the session backends are released?

**Reply by provinzkraut** [reply]
@Goldziher Not a lot really. The session backends are already integrated with `TestClient`. We can now drop `cryptography` from the `testing` extra, since it's not really tied to testing but just the `CookieBackend`, and people can test sessions with other backends as well.

**Reply by Goldziher** [reply]
Ok, can you handle this as part of your session related updates or a separate issue?

**Reply by provinzkraut** [reply]
Will take care of it!

**Comment by Tobi-De** [comment]
I don't know enough about Starlite so my first reaction was: WTF, really?
But after reading the discussion, it totally makes sense, the only concern I have is: wouldn't it be too much work to add right now? you guys are doing a lot, especially you @Goldziher on several repositories at once. I hope this doesn't end in a burn out. Genuine concerne here, not trying to discourage anybody.

Edit: Nevermind, you guys already started 😆

**Reply by Goldziher** [reply]
> I don't know enough about Starlite so my first reaction was: WTF, really?
> But after reading the discussion, it totally makes sense, the only concern I have is: wouldn't it be too much work to add right now? you guys are doing a lot, especially you @Goldziher on several repositories at once. I hope this doesn't end in a burn out. Genuine concerne here, not trying to discourage anybody.
> 
> Edit: Nevermind, you guys already started 😆 

Indeed 😁. Its not immediate - I think it will probably take a month or two to get done.

**Comment by ubernostrum** [comment]
Is the plan to retain compatibility with Starlette middlewares? I've been relying on that to write middlewares that can be reused across multiple Starlette-based projects.

**Reply by Goldziher** [reply]
we will
